jscomp/others/belt_String.ml

```diff
@@ -159,3 +159,88 @@ external substr : from:int -> t = "" [@@bs.send.pipe: t]
 *)
 external trim : t -> t = "" [@@bs.send]
 
+(* link to JS built-ins *)
+external str_length: t -> int = "length" [@@bs.get]
+external str_charAt: t -> int -> t = "charAt" [@@bs.send]
+external array_concat: 'a array -> 'a array -> 'a array = "concat" [@@bs.send]
+
+(**
+  [reduce str f] takes a string argument, initial value, and a reducer
+  function [f] whose arguments are an accumulated value and a character from the string.
+
+  [reduce] starts the accumulated value as the initial value, and updates it by
+  applying the reducer function [f] to each character in the string.
+  
+@example {[
+  let addCharCodes (total: float) (s:string) =
+    (total +. (Js.String.charCodeAt 0 s): float)
+
+  reduce "abc" 0.0 addCharCodes = 294.0
+  
+  let reverser (result: string) (item: string) =
+    (item ^ result : string)
+  
+  reduce "abcde" "" reverser = "edcba"
+]}
+*)
+
+let reduce (s: string) (acc: 'a) (f: 'a -> string -> 'a) =
+  (let rec helper (acc: 'a) (index: int) =
+    match index with
+    | n when n == (str_length s) -> acc
+    | n -> helper (f acc (str_charAt s n)) (n + 1) in
+
+  helper acc 0 : 'a)
+
+(**
+  [stringMap s f] applies a function [f] to each character in the given string [s],
+  returning a new string with the concatenated results of the function calls.
+
+@example {[
+let addDash (s : string) = (s ^ "-" : string)
+stringMap "abcde" addDash = "a-b-c-d-e-"
+]}
+*)
+
+let stringMap (s : string) (f : string -> string) =
+  (reduce s "" (fun acc item -> acc ^ (f item)) : string)
+
+(**
+  [map s f] applies a function [f] to each character in the given string [s],
+  returning an array containing the results of the function calls.
+  
+@example {[
+let toCode (s : string) = (Js.String.charCodeAt 0 s : float)
+map "abcde" toCode = [| 97.0; 98.0; 99.0; 100.0; 101.0 |]
+]}
+*)
+let map (s : string) (f : string -> 'a) =
+  (reduce s [||] (fun acc item -> array_concat acc [|(f item)|]) : 'a array)
```

Comment:
Can you optimize this by allocating the array ahead of time? To make this compatible with native, you should call `Belt_array.makeUninitializeUnsafe` with the length of the array (same as the length of the string) and then call `Belt_array.setUnsafe` :)

Reply:
Good idea. I think I would need to have an index into the string, so I am not sure I could use `reduce` as it currently stands. Also, the length of the string has to take supplementary planes into account, as JavaScript tells me that `"😺".length` is 2. This could get ugly.